### PR TITLE
return the pause result when committing

### PR DESCRIPTION
**Pull request by Mashimiao**
Signed-off-by: Ma Shimiao mashimiao.fnst@cn.fujitsu.com
I think if user used "--pause" when committing, but pause container failed. We should return the errror except container is not  running or is already paused

### CODE REVIEW

**Comment by duglin** [comment]
was trying to think of a reason that we would want to just ignore the error and keep going, but I can't think of one since if the user asked for it to be paused, we should probably error-out if we can't do what they asked.

Not sure how hard this will be, but can you see if you can add a testcase for this?

**Comment by Mashimiao** [comment]
Hi @duglin  Actually, I think this is small probability event and not easy to test. It's related to cgroup freezer.

**Comment by mountkin** [comment]
I think the test can be done with these steps:
1. create a container and start it
2. delete the freezer cgroup of the container
3. commit and check the error

:smile_cat:

**Comment by cpuguy83** [comment]
Other than the inherent race in checking if it is paused and then pausing.

Probably should pause without checking (if it is already paused) and make sure it returns a real typed error that you can check to see if the error is that it is already paused (which itself can then be ignored).

**Comment by crosbymichael** [comment]
What happens on systems that do not have the freezer cgroup now?  They would no longer be able to commit a container or do any builds with this PR.

**Comment by crosbymichael** [comment]
sry, dnd't mean to close

**Comment by cpuguy83** [comment]
@crosbymichael pausing is optional.
I'd expect an error if the is no freezer but the user specified they wanted to pause.

**Comment by Mashimiao** [comment]
Hi @cpuguy83. If a container is paused can be ignored. How about changing container.pause()? If a container is already pasued, just return nil?

**Comment by cpuguy83** [comment]
@Mashimiao I'd rather have `container.Pause` return something like ErrAlreadyPaused and check for that.

**Comment by Mashimiao** [comment]
Hi @cpuguy83, I fix code to just return error which because of freeze.

**Comment by Mashimiao** [comment]
ping @cpuguy83 , @duglin , @crosbymichael

File: daemon/commit.go [hunk 1]
```diff
@@ -20,7 +21,10 @@ type ContainerCommitConfig struct {
 // The image can optionally be tagged into a repository.
 func (daemon *Daemon) Commit(container *Container, c *ContainerCommitConfig) (*image.Image, error) {
 	if c.Pause && !container.isPaused() {
-		container.pause()
+		if err := container.pause(); err != nil && !strings.Contains(err.Error(), "is not running") && strings.Contains(err.Error(), "is already paused") {
```
Comment (jfrazelle) [inline daemon/commit.go]: hmmm i dont know how i feel about relying on these strings to not change, we have had random go lint prs cause strings to change and things like this to break in the past

Comment (cpuguy83) [inline daemon/commit.go]: Agree, we should have an error type and maybe a function like `daemon.isNotRunning(Err)`, and `daemon.isPaused(err)`

Comment (Mashimiao) [inline daemon/commit.go]: How about use error type in `github.com/docker/docker/errors`, like ErrorCodeNotRunning and ErrorCodeAlreadyPaused

File: daemon/commit.go [hunk 2]
```diff
@@ -20,7 +21,10 @@ type ContainerCommitConfig struct {
 // The image can optionally be tagged into a repository.
 func (daemon *Daemon) Commit(container *Container, c *ContainerCommitConfig) (*image.Image, error) {
 	if c.Pause && !container.isPaused() {
-		container.pause()
+		if err := container.pause(); err != nil && err != derr.ErrorCodeNotRunning.WithArgs(container.ID) && err != derr.ErrorCodeAlreadyPaused.WithArgs(container.ID) {
```
Comment (duglin) [inline daemon/commit.go]: I'm not thrilled that there isn't a util func in errcode to make this easier, I would prefer if you could just write something like:
```
if ... derr.ErrorCodeNotRunning.isA(err) ...
```
to test to see if 'err' is of type ErrorCodeNotRunning, because doing a strcmp like this (while probably ok) just feel kind of ugly to me.  But that's not your fault.  If you want to add a util func in the derr package as part of this I wouldn't mind  :-)

But, in the meantime, I think the more proper way to do this is to try to cast 'err' to be of type 'ErrorCoder`, and if it is, then just check the ErrorCode() of err and make sure its the same as `derr.ErrorCodeNotRunning`.  Does that make sense?

Comment (Mashimiao) [inline daemon/commit.go]: I think there is no need to cast 'err', ErrorCodeNotRunning is just a int value. We can just checck err's Code is ErrorCodeNotRunning or not

Comment (Mashimiao) [inline daemon/commit.go]: Yes,  fixed.

**Comment by duglin** [comment]
@Mashimiao I think the errors we're seeing in the tests are real.  Probably related to https://github.com/docker/docker/pull/16005#issuecomment-137543472

**Comment by Mashimiao** [comment]
@duglin  Do I need to fix the test case?

**Comment by duglin** [comment]
@Mashimiao are you able to produce these failures? If so it might be nice to see what error is being generated at your new code so you can look for it and ignore that error too.

While I don't agree with ignoring the error, since the user asked for the container to be paused but we failed, I think changing this to report an error would break too many things. By default `docker commit` will set the pause flag to true, which means I bet a lot of people will now see failures on the commits and I think that's not good. What I don't understand though is why the pause testcases aren't failing.

**Comment by Mashimiao** [comment]
@duglin  all failures are like `Error response from daemon: open /sys/fs/cgroup/freezer/docker/c771e30ddab7732a39808f1f3dde9f68712b5d68571a0142648f6e76b7c73934/freezer.state: no such file or directory`.
I think that's because has a default option `-p, --pause=true    Pause container during commit`.
When committing, try to pause the container but the container is already stopped.

**Comment by duglin** [comment]
@Mashimiao I'm not able to reproduce the issue. If you are, can you add some debug lines to https://github.com/docker/docker/pull/16005/files#diff-73a30faa70ede73d1175e8b5db02b18fR25 to see what err shows up there?

**Comment by Mashimiao** [comment]
Hi @duglin, I tested in my local environment. I can't reproduce this issue either.

**Comment by Mashimiao** [comment]
@duglin From the log of janky and windows, I think janky and windows failed has nothing to do with this PR. It seems the dokcker or test environment is not stable.

**Comment by duglin** [comment]
I think this aligns with what @cpuguy83 was suggesting, so LGTM
/ping @cpuguy83

**Comment by cpuguy83** [comment]
Looks good,I'm going to move to code-review.
Trying to think if there is a way to test this, but may be difficult.

**Comment by cpuguy83** [comment]
Maybe stop a container, and commit with `--pause`?

**Comment by thaJeztah** [comment]
@vdemeester do you have thoughts on how to test this?

File: daemon/commit.go [hunk 3]
```diff
@@ -20,7 +22,18 @@ type ContainerCommitConfig struct {
 // The image can optionally be tagged into a repository.
 func (daemon *Daemon) Commit(container *Container, c *ContainerCommitConfig) (*image.Image, error) {
 	if c.Pause && !container.isPaused() {
-		container.pause()
+		if err := container.pause(); err != nil {
+			switch err.(type) {
+			case errcode.Error:
+				daError, _ := err.(errcode.Error)
+				if daError.ErrorCode() != derr.ErrorCodeNotRunning && daError.ErrorCode() != derr.ErrorCodeAlreadyPaused {
```
Comment (calavera) [inline daemon/commit.go]: I don't think `ErrorCodeNotRunning` and `ErrorCodeAlreadyPaused` are the same. We cannot unpause it in the defer below if the container is not running because it's stopped.

I think we should only skip returning here when the error is `ErrCodeAlreadyPaused`, and we should only try to unpause when it was really paused.

Comment (thaJeztah) [inline daemon/commit.go]: @Mashimiao can you look at this?

Comment (Mashimiao) [inline daemon/commit.go]: Now, we just unpause it in the defer if the container was really paused.
When committing we want to pause container, because we want the container keep unchanged.
So, I think the container wheter is already stopped or paused satisfies our requirement

**Comment by thaJeztah** [comment]
@Mashimiao looks like there are merge-conflicts :crying_cat_face:

**Comment by Mashimiao** [comment]
@thaJeztah  conflict fixed

**Comment by thaJeztah** [comment]
ping @calavera PTAL

@Mashimiao have you had time to see if a test is possible?

**Comment by Mashimiao** [comment]
@thaJeztah   testcase added

**Comment by thaJeztah** [comment]
ping @cpuguy83 @duglin testcase was added PTAL

(thanks @Mashimiao !)

File: integration-cli/docker_cli_commit_test.go [hunk 4]
```diff
@@ -131,8 +147,7 @@ func (s *DockerSuite) TestCommitChange(c *check.C) {
 		"Config.Env":          "[DEBUG=true test=1 PATH=/foo]",
 		"Config.Labels":       "map[foo:bar]",
 		"Config.Cmd":          "{[/bin/sh]}",
-		"Config.WorkingDir":   "/opt",
-		"Config.Entrypoint":   "{[/bin/sh]}",
+		"Config.WorkingDir":   "/opt", "Config.Entrypoint":   "{[/bin/sh]}",
```
Comment (vdemeester) [inline integration-cli/docker_cli_commit_test.go]: I think this change is unrelated right ? :stuck_out_tongue:

File: integration-cli/docker_cli_commit_test.go [hunk 5]
```diff
@@ -37,6 +38,21 @@ func (s *DockerSuite) TestCommitWithoutPause(c *check.C) {
...
 
+func (s *DockerSuite) TestCommitWithoutCgroupFreeze(c *check.C) {
+	testRequires(c, DaemonIsLinux)
+	outId, _ := dockerCmd(c, "run", "-i", "-d", "busybox")
+
+	containerID := strings.TrimSpace(outId)
+	output, _ := dockerCmd(c, "exec", containerID, "cat /proc/self/cgroup | grep freeze | awk -F : '{print $3}'")
+
+	freezePath := strings.TrimSpace(output)
+	os.RemoveAll(strings.Join([]string{"/sys/fs/cgroup/freeze", freezePath}, ""))
```
Comment (mountkin) [inline integration-cli/docker_cli_commit_test.go]: Perhaps the path should be `/sys/fs/cgroup/freezer`?
And a cgroup hierarchy can't be deleted via `rm -rf`, instead we should first move the tasks in the cgroup to its parent hierarchy and then remove the directory via `rmdir`.
FYI http://blog.tinola.com/?e=21

It's weird why the jenkins result isn't showing in the status of this PR.

I think the `dockerCmd` below won't touch the changes that this PR introduced.  It should be `docker commit --pause freezetest` and the command should fail, right?

Comment (cpuguy83) [inline integration-cli/docker_cli_commit_test.go]: This just seems dangerous in the greater scope of things (all the other tests to run).

Comment (mountkin) [inline integration-cli/docker_cli_commit_test.go]: @cpuguy deleting the cgroup of a container won't impact the other tests

Comment (Mashimiao) [inline integration-cli/docker_cli_commit_test.go]: @mountkin --pause default value is true

Comment (mountkin) [inline integration-cli/docker_cli_commit_test.go]: Oh yes, I didn't know that. Thanks.

**Comment by cpuguy83** [comment]
```
FAIL: docker_cli_commit_test.go:41: DockerSuite.TestCommitWithoutCgroupFreezer

docker_cli_commit_test.go:46:
    output, _ := dockerCmd(c, "exec", "--privileged", containerID, "cat /proc/self/cgroup | grep freeze | awk -F : '{print $3}'")
/go/src/github.com/docker/docker/pkg/integration/dockerCmd_utils.go:35:
    c.Assert(err, check.IsNil, check.Commentf("%q failed with errors: %s, %v", strings.Join(args, " "), out, err))
... value *exec.ExitError = &exec.ExitError{ProcessState:(*os.ProcessState)(0xc820393f20)} ("exit status 255")
... "exec --privileged 97ba121f8f14b7378d35d6fb3a481a302f5eed8bca367d086cc767fcdde44351 cat /proc/self/cgroup | grep freeze | awk -F : '{print $3}'" failed with errors: exec: "cat /proc/self/cgroup | grep freeze | awk -F : '{print $3}'": stat cat /proc/self/cgroup | grep freeze | awk -F : '{print $3}': no such file or directory
, exit status 255
```

Also broken on userns since --privileged is not supported there yet.

**Comment by Mashimiao** [comment]
@cpuguy83   The test in my local environment is well, but why in jenkins always occurs `stat cat /proc/self/cgroup | grep freeze | awk -F : '{print $3}': no such file or directory`.
Is there no /proc/self/cgroup file in jenkins container?

**Comment by mountkin** [comment]
Try the following.
The main downside of the my patch is that it uses a hard-coded cgroup path `/sys/fs/cgroup/freezer/docker` :sweat: 
```diff
diff --git a/daemon/commit.go b/daemon/commit.go
index 61808ea..9290355 100644
--- a/daemon/commit.go
+++ b/daemon/commit.go
@@ -4,6 +4,8 @@ import (
 	"fmt"
 	"runtime"
 
+	"github.com/docker/distribution/registry/api/errcode"
+	derr "github.com/docker/docker/errors"
 	"github.com/docker/docker/image"
 	"github.com/docker/docker/pkg/archive"
 	"github.com/docker/docker/pkg/ioutils"
@@ -37,8 +39,14 @@ func (daemon *Daemon) Commit(name string, c *ContainerCommitConfig) (*image.Imag
 	}
 
 	if c.Pause && !container.isPaused() {
-		daemon.containerPause(container)
-		defer daemon.containerUnpause(container)
+		if err := daemon.containerPause(container); err != nil {
+			ec, ok := err.(errcode.Error)
+			if !ok || (ec.ErrorCode() != derr.ErrorCodeNotRunning && ec.ErrorCode() != derr.ErrorCodeAlreadyPaused) {
+				return nil, err
+			}
+		} else {
+			defer daemon.containerUnpause(container)
+		}
 	}
 
 	if c.MergeConfigs {
diff --git a/integration-cli/docker_cli_commit_test.go b/integration-cli/docker_cli_commit_test.go
index 8f650ea..3582f24 100644
--- a/integration-cli/docker_cli_commit_test.go
+++ b/integration-cli/docker_cli_commit_test.go
@@ -1,6 +1,10 @@
 package main
 
 import (
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
 	"strings"
 
 	"github.com/docker/docker/pkg/integration/checker"
@@ -191,3 +195,17 @@ func (s *DockerSuite) TestCommitMergeConfigRun(c *check.C) {
 		c.Fatalf("expected envs to match: %v - %v", config1.Env, config2.Env)
 	}
 }
+
+func (s *DockerSuite) TestCommitWithoutCgroupFreezer(c *check.C) {
+	testRequires(c, DaemonIsLinux)
+	cid, _ := dockerCmd(c, "run", "-d", "busybox", "top")
+	freezerPath := filepath.Join("/sys/fs/cgroup/freezer/docker", strings.TrimSpace(cid))
+	parentPath := filepath.Dir(freezerPath)
+	cmd := fmt.Sprintf("cat %s >> %s", filepath.Join(freezerPath, "tasks"), filepath.Join(parentPath, "tasks"))
+	outBytes, err := exec.Command("sh", "-c", cmd).CombinedOutput()
+	c.Assert(err, checker.IsNil, check.Commentf(string(outBytes)))
+	c.Assert(os.Remove(freezerPath), checker.IsNil)
+
+	out, _, err := dockerCmdWithError("commit", "--pause", "freezetest")
+	c.Assert(err, checker.NotNil, check.Commentf(out))
+}
```

**Comment by Mashimiao** [comment]
@mountkin  not all system's docker default cgroup path is /sys/fs/cgroup/freezer/docker-xxxx. Fox example, in Fedora 22  it's /sys/fs/cgroup/freezer/system.slice/docker-xxxx. So, I think use `cat /proc/self/cgroup` is more suitable.

**Comment by Mashimiao** [comment]
The testcase is always fails. I think that's because the janky is running in a container, we can't cat cgroup files in a container, am I right?

**Comment by thaJeztah** [comment]
ping @cpuguy83 @duglin any suggestions for the testcase?

**Comment by crosbymichael** [comment]
From my original comment this is a change of behavior where commits and build will fail if a system does not have the freezer cgroup.  We did it this way initially because it is the default to pause the container when commiting.  

```
Usage:  docker commit [OPTIONS] CONTAINER [REPOSITORY[:TAG]]

Create a new image from a container's changes
  -p, --pause=true    Pause container during commit

```

I don't think we can do this because ppl will start to have failures on build and commit and not know what changed.

**Comment by Mashimiao** [comment]
The PR has been opened for a long time. As there is not a proper solution to make test by now, how about remove the testcase and add in the future if we found a proper solution?

**Comment by cpuguy83** [comment]
Yeah, it seems we have to ignore freezer errors.

**Comment by Mashimiao** [comment]
@cpuguy83   testcase removed

**Comment by Mashimiao** [comment]
@thaJeztah, @duglin , @cpuguy83  can this PR be merged?

**Comment by duglin** [comment]
@crosbymichael correct me if I'm wrong, but I think you're basically NOT LGTM'ing this PR, correct?

**Comment by thaJeztah** [comment]
ping @crosbymichael can you please answer @duglin's question?

**Comment by thaJeztah** [comment]
ping @crosbymichael, could you please have a look?

**Comment by thaJeztah** [comment]
ping @crosbymichael this PR is stalling, could you please have a look?

**Comment by crosbymichael** [comment]
yes, i don't think we can make this change because it is the default and the reason why we ignored the errors was because not all systems have freezer mounted and should default back to the original behavior if it does not.

**Comment by thaJeztah** [comment]
Thanks for looking @crosbymichael.

Sorry that this didn't get merged @Mashimiao, thanks so much for contributing!

**Comment by Mashimiao** [comment]
@thaJeztah, @crosbymichael   
I don't agree with @crosbymichael 
If a system did not have freezer mounted, pause should be failed.
And we should not user -p option when committing. 
If a system did not have freeze but you used -p=true option, I think you used option in a wrong way.
And if a system did have freezer, with this PR, we can check if it's related with freeze with pause failed.

**Comment by yangdongsheng** [comment]
Well, guys, Let me try to make this problem more clear.
At first, I got the point of @crosbymichael , yes, if the default behaviour of commit is pause == yes, then we should not force user to make sure the system supporting freezer before using docker commit in default manner. The most problem I can think is lots of regression problem and lots of system down in all over the world. 

But Yes, I agree with other guys, if freezer is not supported, why we did not tell user and continue something different? 

What about this idea?
(1) change --pause to false as default in docker commit.
The point is, if pause is not supportted by every system, we should make it optional.
(2) error out if freezer is not supported and user force use with --pause
If user specified the --pause, they have to take responsibility to make the system supporting it. And if not, they need to know.

@Mashimiao @thaJeztah, @crosbymichael

**Comment by thaJeztah** [comment]
@Mashimiao @yangdongsheng Thanks for adding your suggestions / concerns.
This PR was already on our list to be discussed in the next maintainers review session. I'll send a follow-up email that it should be discussed/reviewed, even though it's now closed.

**Comment by thaJeztah** [comment]
@Mashimiao @yangdongsheng we discussed this in the review session and we don't think we should change the current behavior; pausing is not a requirement of commiting, it just helps commiting be more atomic.

The original change was deliberate as freeze is common, but not that common for everyone who uses docker. We have already seen via other issues that even simple changes or cgroup requirements can break people so its better to err on the side of conservative than aggressive with these changes.

**Comment by Mashimiao** [comment]
OK, got it. Thanks @thaJeztah
